sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/V1Beta3Helper.java

```diff
@@ -0,0 +1,486 @@
+    public void startBundle(Context c) throws Exception {
+      datastore = getDatastore(DatastoreFactory.get(), c.getPipelineOptions(),
+          options.getProjectId());
+      querySplitter = DatastoreHelper.getQuerySplitter();
+    }
+
+    @Override
+    public void processElement(ProcessContext c) throws Exception {
+      Query query = c.element();
+
+      // If query has a user set limit, then do not split.
+      if (query.hasLimit()) {
+        c.output(query);
+        return;
+      }
+
+      int estimatedNumSplits;
+      // Compute the estimated numSplits if numSplits is not specified by the user.
+      if (numSplits == 0) {
+        estimatedNumSplits = getEstimatedNumSplits(datastore, query, options.getNamespace());
+      } else {
+        estimatedNumSplits = numSplits;
+      }
+
+      List<Query> querySplits;
+      try {
+        querySplits = splitQuery(query, options.getNamespace(), datastore, querySplitter,
+            estimatedNumSplits);
+      } catch (Exception e) {
+        LOG.warn("Unable to parallelize the given query: {}", query, e);
+        querySplits = ImmutableList.of(query);
+      }
+
+      for (Query subquery : querySplits) {
+        c.output(subquery);
+      }
+    }
+
+    @Override
+    public void populateDisplayData(Builder builder) {
+      super.populateDisplayData(builder);
+      builder
+          .addIfNotNull(DisplayData.item("projectId", options.getProjectId())
+              .withLabel("ProjectId"))
+          .addIfNotNull(DisplayData.item("namespace", options.getNamespace())
+              .withLabel("Namespace"))
+          .addIfNotNull(DisplayData.item("query", options.getQuery().toString())
+              .withLabel("Query"));
+    }
+  }
+
+  /**
+   * A {@link DoFn} that reads entities from datastore for each query.
+   */
+  static class ReadFn extends DoFn<Query, Entity> {
+    private final V1Beta3Options options;
+    // Datastore client
+    private transient Datastore datastore;
+
+    public ReadFn(V1Beta3Options options) {
+      this.options = options;
+    }
+
+    @VisibleForTesting
+    ReadFn(V1Beta3Options options, Datastore datastore) {
+      this.options = options;
+      this.datastore = datastore;
+    }
+
+    @Override
+    public void startBundle(Context c) throws Exception {
+      datastore = getDatastore(DatastoreFactory.get(), c.getPipelineOptions(),
+          options.getProjectId());
+    }
+
+    @Override
+    public void processElement(ProcessContext c) throws Exception {
+      Reader reader = createReader(c.element());
+      if (reader.start()) {
+        do {
+          c.output(reader.getCurrent());
+        } while(reader.advance());
+      }
+    }
+
+    // Creates a new reader for the given query
+    @VisibleForTesting
+    Reader createReader(Query query) {
+      return new Reader(query, options.getNamespace(), datastore);
+    }
+  }
+
+  /**
+   * A {@link org.apache.beam.sdk.io.Source.Reader} style reader to read entities
```

Comment:
Generally agree that you do not need to use the `Reader` API.

I would be okay with nested for loops as long as you still have good test coverage. (Note that if you go this way, you can and should use `DoFnTester` so that you don't have to test the entire transform).

If you separate out an iterator instead, you can test the iterator in isolation -- and need not test the DoFn as long as it's basically `for entity in iterator(query) output entity`.